ISPN-13127 Minimize allocations on common code paths

danberindei opened the pull request:
https://issues.redhat.com/browse/ISPN-13127

pruivo commented on commons/all/src/main/java/org/infinispan/commons/io/ExposedByteArrayOutputStream.java:
```diff
@@ -1,37 +1,44 @@
  */
 @NotThreadSafe
-public final class ExposedByteArrayOutputStream extends ByteArrayOutputStream {
+public final class ExposedByteArrayOutputStream extends OutputStream {
```
this is public API. I don't think we can make this change.
https://docs.jboss.org/infinispan/12.1/apidocs/org/infinispan/commons/io/package-summary.html

danberindei replied:
You're right... I was thought I could get away with it because `ByteArrayOutputStream` just implements `OutputStream`, but it does add some public methods, which means users may declare variables of type `ByteArrayOutputStream`.

I can't delay the allocation if I keep extending `ByteArrayOutputStream`, so I think I'll have to add a new class.

pruivo replied:
the `byte[] buf` from `ByteArrayOutputStream` is available directly. Can't you control how it grows?

danberindei replied:
Copied to `LazyByteArrayOutputStream`

pruivo commented:
should we run perf @danberindei ?

danberindei commented:
> should we run perf @danberindei ?

Yeah, let's make sure I didn't break something :)

Just don't expect big improvements: I just "fixed" stuff I noticed in the profiler while working on ISPN-12941 and ISPN-13124, which do improve things a lot, but only in specific scenarios (max-idle + passivation enabled but all entries fit in memory).

danberindei commented:
run performance tests please

pruivo commented on commons/all/src/main/java/org/infinispan/commons/io/LazyByteArrayOutputStream.java:
```diff
@@ -0,0 +1,131 @@
+ *
+ * @author <a href="mailto://brian.stansberry@jboss.com">Brian Stansberry</a>
+ * @author Dan Berindei
+ * @since 4.0
```
nit
```suggestion
 * @since 13.0
```

pruivo commented on core/src/main/java/org/infinispan/interceptors/impl/PassivationCacheLoaderInterceptor.java:
```diff
@@ -18,47 +18,52 @@
+         activateAfterLoad(key, segment, orderer, activationManager, future, ice, null);
+         return retrievalStage;
+      } else {
+         return retrievalStage.whenComplete((value, t) -> {
+            activateAfterLoad(key, segment, orderer, activationManager, future, value, t);
+         });
```
nit:
```suggestion
         return retrievalStage.whenComplete((value, t) -> activateAfterLoad(key, segment, orderer, activationManager, future, value, t));
```

danberindei replied:
I don't like that formatting, the line becomes way too long.

pruivo commented on core/src/main/java/org/infinispan/interceptors/impl/PassivationClusteredCacheLoaderInterceptor.java:
```diff
@@ -1,24 +1,40 @@
                                                                                 int segment, FlagAffectedCommand cmd) {
-      Supplier<CompletionStage<InternalCacheEntry<K, V>>> supplier = () -> super.loadAndStoreInDataContainer(ctx, key, segment, cmd);
-      return PassivationCacheLoaderInterceptor.handlePassivationLoad(key, segment, orderer, activationManager, supplier);
+       CompletableFuture<DataOperationOrderer.Operation> future = new CompletableFuture<>();
```
nit
```suggestion
      CompletableFuture<DataOperationOrderer.Operation> future = new CompletableFuture<>();
```

pruivo commented on core/src/main/java/org/infinispan/interceptors/impl/PassivationClusteredCacheLoaderInterceptor.java:
```diff
@@ -1,24 +1,40 @@
+           PassivationCacheLoaderInterceptor.activateAfterLoad(key, segment, orderer, activationManager, future, ice, null);
+           return retrievalStage;
+       } else {
+           return retrievalStage.whenComplete((value, t) -> {
+               PassivationCacheLoaderInterceptor.activateAfterLoad(key, segment, orderer, activationManager, future, value, t);
+           });
```
nit
```suggestion
           return retrievalStage.whenComplete((value, t) -> activateAfterLoad(key, segment, orderer, activationManager, future, value, t));
```

also, you can import the static method `import static org.infinispan.interceptors.impl.PassivationCacheLoaderInterceptor.activateAfterLoad;` to avoid writing the class name and make the lines shorter :)

danberindei replied:
I can do the static import, but I think the line is still too long even without it

pruivo commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:
```diff
@@ -292,39 +294,59 @@ private boolean allStoresSegmentedOrShared() {
+         if (unavailableExceptionMessage == null) {
+            log.debugf("Store %s is unavailable!", unavailableStore);
+            unavailableExceptionMessage = "Store " + unavailableStore + " is unavailable";
+            CompletionStage<Void> notificationStage = cacheNotifier.notifyPersistenceAvailabilityChanged(false);
+            return notificationStage;
```
nit
```suggestion
            return cacheNotifier.notifyPersistenceAvailabilityChanged(false);
```

pruivo commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:
```diff
@@ -670,46 +692,70 @@ private boolean containedInAdapter(NonBlockingStore nonBlockingStore, String ada
+                && predicate.test(storeStatus.config)) {
+               stageBuilder.dependsOn(storeStatus.store.delete(segment, key)
+                                     .thenAccept(removed -> {
+                                        // OR
```
or?

danberindei replied:
I see now how it's hurting comprehension more than it's helping :)

pruivo commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:
```diff
@@ -670,46 +692,70 @@ private boolean containedInAdapter(NonBlockingStore nonBlockingStore, String ada
-            },
-            this::releaseReadLock
-      ).toCompletionStage();
+      AtomicBoolean removedAny = new AtomicBoolean();
```
wouldn't be better to have a custom `AggregateCompletionStage` that handles the logic of merging the boolean values? 
Something generic like it could do "and" or "or" with all the receives values.

danberindei replied:
A reducer function would work, but I didn't want to stray too much from the "micro-optimization" path. Let me give it a try.

danberindei replied:
I don't think it would work as a AggregateCompletionStage, because of `dependsOn(CompletionStage<?>).

The package doesn't have `@public` so it's not included in the javadoc, but I would rather merge this PR with only minor ("micro") changes.

pruivo commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:
```diff
@@ -766,49 +813,73 @@ private boolean containedInAdapter(NonBlockingStore nonBlockingStore, String ada
-            },
-            this::releaseReadLock);
+                     if (!characteristics.contains(Characteristic.SEGMENTABLE) &&
+                                        !segments.containsAll(IntSets.immutableRangeSet(segmentCount))) {filterToUse = PersistenceUtil.combinePredicate(segments, keyPartitioner, filter);
```
some weird formating here :)

pruivo commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:
```diff
@@ -766,49 +813,73 @@ private boolean containedInAdapter(NonBlockingStore nonBlockingStore, String ada
+   }
+
+   private <K, V> CompletionStage<MarshallableEntry<K, V>>
+   loadFromStoresIterator(Object key, int segment, Iterator<StoreStatus> iterator, boolean localInvocation, boolean includeStores) {
```
is it intentional the method name be on a new line?

danberindei replied:
Yes, I'm trying to keep lines under 120 characters and align parameter names when splitting lines, and long return types don't make it easy :)

pruivo reviewed (changes requested):
LGTM!
just some small nitpicks here and there :)

danberindei commented on commons/all/src/main/java/org/infinispan/commons/io/LazyByteArrayOutputStream.java:
```diff
@@ -0,0 +1,131 @@
+      if (buf == null) {
+         // Pretend we have half the default size so it's doubled
+         buf = new byte[Math.max(DEFAULT_SIZE, newcount)];
+      }
```
Ok, I'll change the next if to `else if`

danberindei commented on core/src/main/java/org/infinispan/interceptors/impl/CacheLoaderInterceptor.java:
```diff
@@ -349,46 +349,52 @@ protected boolean canLoad(Object key) {
     */
-   protected CompletionStage<Void> loadInContext(InvocationContext ctx, Object key, FlagAffectedCommand cmd) {
-
+   protected CompletionStage<?> loadInContext(InvocationContext ctx, Object key, int segment, FlagAffectedCommand cmd) {
```
Did you mean `CompletionStage<Void>`?

It's the future that gets stored in the `pendingLoads` map is a `CompletableFuture<InternalCacheEntry<K, V>>`, and I wanted to be able to return that future without a `thenAccept` call (allocating another lambda) or a hacky cast to an erased type.

pruivo replied:
the `CompletableFuture` from this method is never stored in the `pendingLoads`. 
It is consumed by `asyncInvokeNext()` method (at least, IDEA is unable to find any other usages)

danberindei replied:
@pruivo see https://github.com/infinispan/infinispan/blob/2dc0d28be7cbadaf414cdc0eb48d2d3031e8eef7/core/src/main/java/org/infinispan/interceptors/impl/CacheLoaderInterceptor.java#L358

We put `cf` in the `pendingLoads` map and we also return it from the method.

pruivo replied:
I was definitely blind xD

danberindei commented on core/src/main/java/org/infinispan/interceptors/impl/CacheLoaderInterceptor.java:
```diff
@@ -349,46 +349,52 @@ protected boolean canLoad(Object key) {
-      });
-      return cf.thenAccept(action);
+      if (CompletionStages.isCompletedSuccessfully(result)) {
+         finishLoadInContext(ctx, key, cmd, cf, CompletionStages.join(result), null);
```
`cf` is always used:

https://github.com/infinispan/infinispan/pull/9408/files/0c4418d7283a5b76495801b7e69ebbe10b33677f#diff-90e48286ca570ac9d195f046f4dc37c59507eb0fb7e847fb80040ad79d94e88bR359

pruivo replied:
link not working :(

danberindei replied:
Sorry, I rebased and force-pushed!
I put the proper link in the discussion above, let's merge the threads there.

danberindei commented:
@pruivo I've addressed most of your nits (some I don't agree with, one I feel is too much work for this PR).

I've also "future-proofed" the `PersistenceManagerImpl.deleteFromAllStores()` code for https://github.com/infinispan/infinispan/pull/9471

danberindei commented:
I rebased again to merge the review fixes into the original commits.

pruivo commented:
@danberindei there are some tests failing with NullPointerException. It is not clear if it is related.

danberindei commented:
@pruivo the NPE was related. I added a commit to fix the exception handling in `PersistenceManagerImpl.deleteFromAllStores`

pruivo commented:
looks like there is another: https://ci.infinispan.org/job/Infinispan/job/PR-9408/9/testReport/org.infinispan.query.remote.impl/ProtobufMetadataManagerInterceptorTest/testStatusAfterRemove/
